### PR TITLE
[Merged by Bors] - feat: presheaf of modules over a presheaf of rings

### PR DESCRIPTION
**Pull request by kim-em**
This is extracted from the draft PR https://github.com/leanprover-community/mathlib4/pull/4116 which tries to compare this definition with the definition in terms of a presheaf in `RingMod`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Category/ModuleCat/Presheaf.lean [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+  map_add' x y := rfl
+  map_smul' r x := by
+    have := (RingHomId.eq_id : σ = _)
+    subst this
+    rfl
+
+open LinearMap
+
+open Opposite
+
```
Comment (kbuzzard) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf.lean:51]: Are we now in a post "declare universe v1 before u1" world or is this an oversight?

**Review by kbuzzard — APPROVED** [review]
This looks great! We have to start somewhere and I propose we start here.

Note that this is a (much much better,  but developed independently as far as I know) version of the experimental file https://github.com/ImperialCollegeLondon/tcc-lean-alg-geom-2022/blob/master/src/sheaves_of_modules/defs.lean which I successfully used to do some more very basic stuff in an 8 hour graduate course; I only cite this repo because it can be considered as evidence that the approach in this PR is viable.

File: Mathlib/Algebra/Category/ModuleCat/Presheaf.lean [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+/-!
+# Presheaves of modules over a presheaf of rings.
+
+We give a hands-on description of a presheaf of modules over a fixed presheaf of rings,
+as a presheaf of abelian groups with additional data.
+
+## Future work
+
+* Show that we can check the sheaf condition at the level of abelian groups.
```
[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Algebra/Category/ModuleCat/Presheaf.lean:17]: Indeed, this was me blindly copy-pasting from an earlier draft about RingMod.

**Review by jcommelin — APPROVED** [review]
Let's get this merged! The proof of the pudding... can't be established without the eating.

bors merge

**Pull request closed without merging**